### ISSUE TITLE
Refactor DNSProtocol and DNSDatagramProtocol

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2414]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)| @therve reported|
|-|-|
|Trac ID|trac#2414|
|Type|enhancement|
|Created|2007-01-21 17:26:35Z|

DNSProtocol and DNSDatagramProtocol should have a common base class: some things have been corrected in DNSDatagramProtocol and not in DNSProtocol because of the split (like the use of Deferred.setTimeout).

Also, tests must been added to DNSProtocol.

<details><summary>Searchable metadata</summary>

```
trac-id__2414 2414
type__enhancement enhancement
reporter__therve therve
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__names names
keywords__ 
time__1169400395000000 1169400395000000
changetime__1174822934000000 1174822934000000
version__None None
owner__ 
cc__therve cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_1"></a>|@therve removed owner|
|-|-|

Ready to review in names-dnsprotocol-2414.

I didn't make too many changes, because I don't want the branch to be huge, and I don't master twisted.names. I've just refactored DNSDatagramProtocol and DNSProtocol, added a few tests, and a few docstrings.

Any comment is welcome.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

I'm not done reviewing/thinking about this branch.  Some questions I am considering:

 * is a DNSProtocol still useful after a query has timed out?
 * Should the default DNSProtocol.query have a timeout?  If so, it should probably be 60, not 10.
 * Is there sufficient test coverage for the timeout case for DNSProtocol?

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve commented|
|-|-|

Replying to [exarkun](#note_2):
> I'm not done reviewing/thinking about this branch.  Some questions I am considering:
>
>  * is a DNSProtocol still useful after a query has timed out?

That's actually a good question :). It's a application timeout, so the connection might me always usable, no ?

>  * Should the default DNSProtocol.query have a timeout?  If so, it should probably be 60, not 10.

10 was the old default timeout, so I just kept it.

>  * Is there sufficient test coverage for the timeout case for DNSProtocol?

If you ask it, he anwer is probably no :).

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve commented|
|-|-|

I've added a test statement in the timeout case, and changed default timeout to 60.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph set owner to @exarkun|
|-|-|

I suspect that the only person competent to review this change is exarkun.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @therve|
|-|-|

Okay, keeping the old timeout is good.  I mistakenly thought there was no timeout previously.

Some of the test methods do things with Deferreds but don't return them because they complete synchronously.  Still, they should return the Deferreds to trial so that if somehow the Deferreds haven't fired by the time the test method is over, trial will notice this and fail the test.  Otherwise the tests will pass even though the code isn't working properly.

Specifically, DatagramProtocolTestCase.test_queryTimeout and DNSProtocolTestCase.test_queryTimeout

Other than this, this branch looks good to merge.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve commented|
|-|-|

To clarify, there was no timeout before, my previous comment was wrong. It was 10 for UDP so I just copied it. In the last version I switched to 60.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

Hmm :)

Okay, that still seems fine.  In practice, when using the higher-level APIs that are mostly used, these low level timeouts are always overridden.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve set status to `closed`|
|-|-|

(In [19905]) Merge names-dnsprotocol-2414

Author: therve
Reviewer: exarkun
Fixes [#2414](https://github.com/twisted/twisted/issues/2414)

Refactor DNSDatagramProtocoland DNSProtocol to use the same base class: it
removes some of the redundancy, and in particular a setTimeout call. Also, it
improves test coverage of the touched code.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
